cabal-validate/main/Main.hs

```diff
@@ -0,0 +1,955 @@
+
+      tPutStrLn $
+        tSetSGRCode [SetColor Foreground Vivid Green]
+          <> "Finished after "
+          <> formatDiffTime duration
+          <> ": "
+          <> prettyCommand
+          <> "\nTotal time so far: "
+          <> formatDiffTime totalDuration
+          <> tSetSGRCode [Reset]
+    ExitFailure exitCode' -> do
+      unless opts.verbose $ do
+        tPutStrLn output
+
+      tPutStrLn $
+        tSetSGRCode [SetColor Foreground Vivid Red]
+          <> "Failed with exit code "
+          <> tShow exitCode'
+          <> " after "
+          <> formatDiffTime duration
+          <> ": "
+          <> prettyCommand
+          <> "\nTotal time so far: "
+          <> formatDiffTime totalDuration
+          <> tSetSGRCode [Reset]
+
+      throwIO
+        ProcFailed
+          { procCommand = command
+          , procArguments = args
+          , procExitCode = exitCode
+          }
+
+getAbsoluteTime :: IO AbsoluteTime
+getAbsoluteTime = systemToTAITime <$> getSystemTime
+
+formatDiffTime :: DiffTime -> Text
+formatDiffTime delta =
+  let minute = secondsToDiffTime 60
+      hour = 60 * minute
+   in T.pack $
+        if delta >= hour
+          then formatTime defaultTimeLocale "%h:%02M:%02ES" delta
+          else
+            if delta >= minute
+              then formatTime defaultTimeLocale "%m:%2ES" delta
+              else formatTime defaultTimeLocale "%2Ess" delta
+
+main :: IO ()
+main = do
+  opts <- execParser fullOptsParser
+  resolvedOpts <- resolveOpts opts
+  mainInner resolvedOpts
+
+mainInner :: ResolvedOpts -> IO ()
+mainInner opts =
+  forM_ opts.steps $ \step -> do
+    runStep opts step
+
+printConfig :: ResolvedOpts -> IO ()
+printConfig opts = do
+  tPutStrLn $
+    "compiler:          "
+      <> T.pack opts.compiler.executable
+      <> "\ncabal-install:     "
+      <> T.pack opts.cabal
+      <> "\njobs:              "
+      <> tShow opts.jobs
+      <> "\nsteps:             "
+      <> T.unwords (map displayStep opts.steps)
+      <> "\nHackage tests:     "
+      <> tShow opts.hackageTests
+      <> "\nverbose:           "
+      <> tShow opts.verbose
+      <> "\nextra compilers:   "
+      <> T.pack (unwords opts.extraCompilers)
+      <> "\nextra RTS options: "
+      <> T.unwords (rtsArgs opts)
+
+printToolVersions :: ResolvedOpts -> IO ()
+printToolVersions opts = do
+  timed opts (T.pack opts.compiler.executable) ["--version"]
+  timed opts (T.pack opts.cabal) ["--version"]
+
+  forM_ opts.extraCompilers $ \compiler -> do
+    timed opts (T.pack compiler) ["--version"]
+
+build :: ResolvedOpts -> IO ()
+build opts = do
+  printHeader "build (dry run)"
+  timed
+    opts
+    (T.pack opts.cabal)
+    ( cabalNewBuildArgs opts
+        ++ opts.targets
+        ++ ["--dry-run"]
+    )
+
+  printHeader "build (full build plan; cached and to-be-built dependencies)"
+  timed
+    opts
+    "jq"
+    [ "-r"
+    , -- TODO: Aeson...? Or even just make this a Cabal feature!
```

Comment:
I think you're looking for [`cabal-plan`](https://hackage.haskell.org/package/cabal-plan). At some point this should be made an external command (will require some option processing changes, see https://github.com/haskell/cabal/issues/10275). I don't know if it's worth integrating into `cabal-install` itself, especially given the dependency footprint of `aeson`.